internal/controller/storageclient_controller.go

```diff
@@ -55,54 +64,143 @@ const (
+				vgsc.Status.VolumeGroupSnapshotHandle != nil {
+				parts := strings.Split(*vgsc.Status.VolumeGroupSnapshotHandle, "-")
+				if len(parts) == 9 {
+					// second entry in the volumeID is clusterID which is unique across the cluster
+					return []string{parts[2]}
+				}
+			}
+			return nil
+		}); err != nil {
+			return fmt.Errorf("unable to set up FieldIndexer for VGSC csi driver name: %v", err)
+		}
+	}
+
+	if err := mgr.GetCache().IndexField(ctx, &csiopv1a1.ClientProfile{}, ownerUIDIndexName, func(obj client.Object) []string {
 		refs := obj.GetOwnerReferences()
-		var owners []string
+		owners := []string{}
 		for i := range refs {
 			owners = append(owners, string(refs[i].UID))
 		}
 		return owners
 	}); err != nil {
-		return fmt.Errorf("unable to set up FieldIndexer for StorageClaim's owner uid: %v", err)
+		return fmt.Errorf("unable to set up FieldIndexer for client profile owner: %v", err)
 	}
 
+	enqueueStorageClientRequest := handler.EnqueueRequestsFromMapFunc(
+		func(ctx context.Context, _ client.Object) []reconcile.Request {
+			storageClients := &v1alpha1.StorageClientList{}
+			if err := r.Client.List(ctx, storageClients); err != nil {
+				return []reconcile.Request{}
+			}
+
+			request := []reconcile.Request{}
+			for idx := range storageClients.Items {
+				request = append(request, reconcile.Request{NamespacedName: client.ObjectKeyFromObject(&storageClients.Items[idx])})
+			}
+			return request
+		},
+	)
+
 	r.recorder = utils.NewEventReporter(mgr.GetEventRecorderFor("controller_storageclient"))
 	generationChangePredicate := predicate.GenerationChangedPredicate{}
 	bldr := ctrl.NewControllerManagedBy(mgr).
 		For(&v1alpha1.StorageClient{}).
-		Owns(&v1alpha1.StorageClaim{}).
 		Owns(&batchv1.CronJob{}).
 		Owns(&quotav1.ClusterResourceQuota{}, builder.WithPredicates(generationChangePredicate)).
 		Owns(&nbv1.NooBaa{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
 		Owns(&corev1.Secret{}).
 		Owns(&csiopv1a1.CephConnection{}, builder.WithPredicates(generationChangePredicate)).
-		Owns(&csiopv1a1.ClientProfileMapping{}, builder.WithPredicates(generationChangePredicate))
+		Owns(&csiopv1a1.ClientProfileMapping{}, builder.WithPredicates(generationChangePredicate)).
+		Owns(&storagev1.StorageClass{}).
+		Owns(&snapapi.VolumeSnapshotClass{}).
+		Owns(&replicationv1a1.VolumeReplicationClass{}, builder.WithPredicates(generationChangePredicate)).
+		Owns(&csiopv1a1.ClientProfile{}, builder.WithPredicates(generationChangePredicate)).
+		Watches(
+			&extv1.CustomResourceDefinition{},
+			enqueueStorageClientRequest,
+			builder.WithPredicates(
+				utils.NamePredicate(VolumeGroupSnapshotClassCrdName),
+				utils.EventTypePredicate(
+					!r.AvailableCrds[VolumeGroupSnapshotClassCrdName],
+					false,
+					true,
+					false,
+				),
+			),
+			builder.OnlyMetadata,
+		)
+	if r.AvailableCrds[VolumeGroupSnapshotClassCrdName] {
+		bldr = bldr.Owns(&groupsnapapi.VolumeGroupSnapshotClass{})
+	}
```

Comment:
We know how to add events dynamically. Why are we moving code that contain a legacy pattern?

Reply:
pinged for ack to make changes offline.

Reply:
will be worked in a different PR.